datafusion/physical-plan/src/sorts/batches.rs

```diff
@@ -0,0 +1,124 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use super::cursor::{Cursor, CursorValues};
+
+/// A representation of a record batch,
+/// passable and sliceable through multiple merge nodes
+/// in a cascaded merge tree.
+///
+/// A `BatchCursor` encapsulates the ability to sort merge each
+/// sliced portition of a record batch, with minimal overhead.
+///
+/// ```text
+/// ┌────────────────────────┐
+/// │ CursorValues   Batch   │           ┌──────────────────────┐
+/// │ ┌──────────┐ ┌───────┐ │    ─ ─ ─ ▶│      BatchTracker    │
+/// │ │  1..10   │ │   A   │ ┼ ─ │       └──────────────────────┘
+/// │ ├──────────┤ ├───────┤ │   │            Holds batches
+/// │ │  11..20  │ │   B   │ ┼ ─ ┘          and assigns BatchId
+/// │ └──────────┘ └───────┘ │
+/// └────────────────────────┘
+///             │
+///             │
+///             ▼
+///        BatchCursors
+/// ┌────────────────────────┐           ┌──────────────────────┐ ─ ▶ push batch
+/// │    Cursor     BatchId  │    ─ ─ ─ ▶│   LoserTree (Merge)  │ ─ ▶ advance cursor
+/// │ ┌──────────┐ ┌───────┐ │   │       └──────────────────────┘ ─ ▶ push row
+/// │ │  1..10   │ │   A   │ ┼ ─ │       ┌──────────────────────┐         │
+/// │ ├──────────┤ ├───────┤ │   │       │   SortOrderBuilder  ◀┼ ─ ─ ─ ─ ┘
+/// │ │  11..20  │ │   B   │ ┼ ─ ┘       └──────────────────────┘
+/// │ └──────────┘ └───────┘ │                holds sorted rows
+/// └────────────────────────┘              up to ceiling size N
+///                                                  │
+///              ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘
+///             │
+///             ▼
+///        BatchCursors
```

Comment:
I'm not sure of a better name, but to me `BatchCursor` would suggest this construction owns a `RecordBatch` and is a cursor over it, which isn't the case